Remove feedback request action for product variations

iamgabrielma opened the pull request:
Closes: #6682 

### Description

We currently have a banner on the Products tab announcing the product variations feature. That feature is no longer new, so it can be removed ( p1650475217179219-slack-CGPNUU63E ).

This PR removes the `productsVariations` as FeedbackType, removes the feedback banner from the Product View, as well as   updates Unit Tests to use `FeedbackType.ordersCreation` rather than `FeedbackType.productsVariations` when testing visibility use-cases and store settings, as we can still keep these tests and are not based on products variations specifically.

It also adds a `FeedbackType.none` case, so don't always necessarily have to pass a feedback case. 

### Question:

Do we want to clear `case productsVariationsFeedback` as well from `SurveyViewController`? This seems to involve configuring and tracking surveys completion for these features. I understand we do not need it anymore for products variations, but I wanted to double-check first as involves tracking and analytics. Created a new issue here to keep the scope of this PR narrowed down: https://github.com/woocommerce/woocommerce-ios/issues/7139


### Testing instructions
When navigating to the Product tab, note that the banner announcing the product variations feature does not appear anymore.

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

iamgabrielma committed:
removed feedback request action for product variations

iamgabrielma committed:
Removed productsVariations FeedbackType

iamgabrielma committed:
Update tests from productsVariation to ordersCreation FeedbackType

rachelmcr commented:
> It also adds a FeedbackType.none case, so don't always necessarily have to pass a feedback case.

I'm a bit worried about this change, because the `AppSettingsAction.updateFeedbackStatus` action relies on the `FeedbackType` to make sure the new status is set on the relevant type of feedback. (For example, to prevent a particular banner from being shown again after it is dismissed.) Since there isn't a scenario in which we want to update the feedback status without specifying the feedback type (that I'm aware of) this runs the risk of misusing that action by using the `none` case when a type needs to be specified.

I might be missing the point, though; what is the problem you're solving with this new case?

iamgabrielma commented:
> I'm a bit worried about this change, because the `AppSettingsAction.updateFeedbackStatus` action relies on the `FeedbackType` to make sure the new status is set on the relevant type of feedback. (For example, to prevent a particular banner from being shown again after it is dismissed.) Since there isn't a scenario in which we want to update the feedback status without specifying the feedback type (that I'm aware of) this runs the risk of misusing that action by using the `none` case when a type needs to be specified.

> I might be missing the point, though; what is the problem you're solving with this new case?

Thanks for the comment! The idea is that, because of the `AppSettingsAction.updateFeedbackStatus` action we always need to pass a `FeedbackType`, but functions like `dismissProductsBanner()` ( and by extension `requestAndShowNewTopBannerView()`) seem to rely on it, by not providing a `.none` case we always need to provide a specific case.

My intention was to keep this generic, so we can reuse the existing logic if we need to use these banners for a different announcement within the Product view in the future, by passing a `FeedbackType.none` case, and [not updating the feedback status](https://github.com/woocommerce/woocommerce-ios/pull/7129/files#diff-0e99eca665e7650a698e692a7a5f42c3910fe9a2eb8de265a0231f1789af6496R42-R43). As at the moment, the ProductsViewController seem to have a bunch of functions that are entirely based on this banner and there's no fallback to just keep the code but not have a banner.

This was my understanding anyway! It's true that I may be missing some context about the usage of updating the feedback status. What would you recommend? Another option could be to remove the `FeedbackType.none` case in order to avoid potential side effects, then remove the banner logic entirely from the ProductsViewController.

rachelmcr commented:
> My intention was to keep this generic, so we can reuse the existing logic if we need to use these banners for a different announcement within the Product view in the future, by passing a FeedbackType.none case, and [not updating the feedback status](https://github.com/woocommerce/woocommerce-ios/pull/7129/files#diff-0e99eca665e7650a698e692a7a5f42c3910fe9a2eb8de265a0231f1789af6496R42-R43).

Thanks, I see what you mean. However, I still think this approach carries the risk that we won't remember to update the case everywhere that needs it and the `none` case could get left when it's not intended. That seems especially risky if we set the `none` case as a default value so it's not required when you call that action:

```swift
case updateFeedbackStatus(type: FeedbackType? = FeedbackType.none, status: FeedbackSettings.Status, onCompletion: ((Result<Void, Error>) -> Void))
```

If we want to keep it generic/reusable, I think it might make sense to use an optional property (e.g. a `FeedbackType` property) on the VC that sets the feedback type. Then, anywhere the feedback type is used (e.g. to dismiss the banner) we can use a `guard` statement to check that the property is set. That way, when we want to remove a banner from Products we can leave that property `nil` and when we want to use the banner for something new we can set that property to its new value.

That said, it may not be worth it to make those changes. As you mentioned, we could just remove the code — since we may want to show a different banner there again in the future, we could just comment out the code related to showing/hiding the banner so it's available the next time we want to use it?

iamgabrielma committed:
Reverted FeedbackType changes. Removed dismissProductsBanner()

iamgabrielma commented:
> This approach carries the risk that we won't remember to update the case everywhere that needs it and the none case could get left when it's not intended. That seems especially risky if we set the none case as a default value so it's not required when you call that action

Thanks for the pointers @rachelmcr, that definitely makes sense. 

I went ahead and reverted the previous changes I made to `FeedbackType`, and moved ahead with the option of removing the minimum amount of code. By taking out `dismissProductsBanner()` we don't need to check for `FeedbackType`, and we can keep most of the logic that is used to show banners in the Product view. We can always rewrite the dismiss function if we need it in the future, or restore it back from git history.

rachelmcr commented on WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:
```diff
@@ -439,21 +439,6 @@ private extension ProductsViewController {
             requestAndShowErrorTopBannerView()
             return
```
Since this is now the only thing this method does, what do you think of removing the `guard` statement (and updating the method documentation to reflect what it does now)? This could be a regular check like:

```swift
    /// Displays an error banner if there is an error loading products data.
    ///
    func showTopBannerViewIfNeeded() {
        if hasErrorLoadingData {
            requestAndShowErrorTopBannerView()
        }
    }
```

rachelmcr commented on WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:
```diff
@@ -466,8 +451,7 @@ private extension ProductsViewController {
             self?.updateTableHeaderViewHeight()
         }, onGiveFeedbackButtonPressed: { [weak self] in
             self?.presentProductsFeedback()
-        }, onDismissButtonPressed: { [weak self] in
-            self?.dismissProductsBanner()
```
Since we're leaving the code to present this banner, how about replacing this with a method that still dismisses the banner? We could use `hideTopBannerView()` to dismiss the banner without making any calls to the `updateFeedbackStatus` action.

iamgabrielma replied:
Thanks for the pointer, I didn't want to create a new dismiss function to not use it, but I didn't see we can use the existing `hideTopBannerView()` for this.

rachelmcr reviewed (commented):
Thanks for making these changes! I left a couple suggestions below.

Since you're only removing the code around checking the feedback visibility and updating the feedback status, but leaving the method that presents the banner, what do you think about leaving some comments as tips/reminders for when we decide to use the banner again? Something that points to those feedback visibility/status actions so we don't forget to add them back. Otherwise I could imagine having the same issue we saw on the Orders tab in https://github.com/woocommerce/woocommerce-ios/issues/6701.

iamgabrielma committed:
Add dismiss func. Add clarification comment

rachelmcr commented on WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:
```diff
@@ -431,32 +431,19 @@ private extension ProductsViewController {
-        guard !hasErrorLoadingData else {
+        if hasErrorLoadingData {
             requestAndShowErrorTopBannerView()
             return
```
tiny nit: We don't need an explicit `return` here anymore.

rachelmcr reviewed (approved):
Looks great! Thanks for taking the time to talk through this with me. 🚀

iamgabrielma committed:
Removed unnecessary return
